### ISSUE TITLE
How to get outputs from an intermediate layer?

### ISSUE THREAD

**Issue by paganpasta** [patrick-kidger/equinox#186]
Hi @patrick-kidger,

I was looking into obtaining outputs from some specific layer of the network without having to modify the network `__call__` method or the network itself. 
 
One way I think of achieving this is using `id_tap` from `jax.experimental.hcb...`. The solution would be to wrap the target `layer` of the network in a class with integrated `id_tap`. Something like `net = eqx.tree_at(lambda net: net.layer_0, net, Wrapper(net.layer_0))`. 

1. Would this be the best way to capture intermediate outputs?
2. A network without BN seems to be working fine (prints normally with `id_print`). However, with `BatchNorm`, throws an error 
```bash
>     flat_results = outside_call_p.bind(*flat_args, **params)
E     TypeError: _outside_call_batching_rule() missing 1 required keyword-only argument: 'result_treedef'

...python3.8/site-packages/jax/experimental/host_callback.py:763: TypeError
```

**Comment by patrick-kidger** [comment]
I think there's a more efficient way to capture intermediate outputs:

```python
def make_wrapper():
    ref = []

    class Wrapper(eqx.Module):
        layer: eqx.nn.Module

        def __call__(self, *args, **kwargs):
            out = self.layer(*args, **kwargs)
            ref.append(out)
            return out

    return ref, Wrapper

def wrap_model(model):
    ref1, wrapper1 = make_wrapper()
    ref2, wrapper2 = make_wrapper()

    model = eqx.tree_at(lambda m: m.layers[1], model, wrapper1(m.layers[1]))
    model = eqx.tree_at(lambda m: m.layers[2], model, wrapper2(m.layers[2]))

    class Model(eqx.Module):
        model: eqx.Module

        def __call__(self, *args, **kwargs):
            out = self.model(*args, **kwargs)
            [intermediate1] = ref1
            [intermediate2] = ref2
            intermediate = (intermediate1, intermediate2)
            return out, intermediate

    return Model(model)

mlp = eqx.nn.MLP(...)

@eqx.filter_jit
def make_step(model, data):
    model = wrap_model(model)
    out, intermediate = model(data)
    return out, intermediate

make_step(mlp, ...)
```

Rationale:
- `host_callback` is generally a performance hit, which this manages to avoid.
- We assume that each intermediate layer is called precisely once, so that `ref` ends up being a list precisely one element long. This isn't mandatory: unpack `ref1` and `ref2` into multiple elements if needed; or just return `ref1` and `ref2` directly.
- Note that both `Wrapper` and `Model` are PyTrees, so that all of the leaves of the original `model` will be visible to e.g. any `jax.grad` transformations you choose to add.
- Whilst this might look suspiciously like a side-effect, I don't think it should count as one so far as JAX is concerned. e.g. if you were to write `out, intermediate = jax.vmap(model)(data)` instead - and thus introduce another level of tracers - then the `intermediate` tracers that have been saved will still pass through the `vmap` on their way out; they're never smuggled out.

**Comment by paganpasta** [comment]
Thanks for the solution!  Yes, it works with vmap'ing as well.

I was working on accessing forward and backward values specific to a layer. Dropping `callbacks`, easier solution to pick up gradients would be to simply access the required field from the output of `filter_{grad,*}` with `output.layer.weight...`.

**Comment by patrick-kidger** [comment]
FWIW accessing forward/backward intermediate values is one of the many things that should become easier once stateful operations make their way into JAX core.

**Comment by paganpasta** [comment]
Regarding stateful, at the moment, with BatchNorm the gradients are not available for layers in a network if its in inference mode. It seems to be intended. 
For this case, maybe the best solution is to replace the BatchNorm layers with FrozenBatchNorm which applies weight and bias without tracking stats.

**Comment by patrick-kidger** [comment]
I'm not sure what you mean. Can you expand? `BatchNorm` should be applying weight and bias regardless of the inference-mode toggle:

https://github.com/patrick-kidger/equinox/blob/18d260d4abab92b6853f2bdbd8335fe2ec316416/equinox/experimental/batch_norm.py#L183

**Comment by paganpasta** [comment]
BatchNorm does apply the weights and bias as you mentioned. 

For computing gradients of a deserialised model in inference mode the error I get is below. Only happens when there is a BatchNorm in the model. 
 
```python

            flat, treedef = jtu.tree_flatten((state, like))
            out = _batchify_p.bind(
                *flat,
                treedef=treedef,
                like_batch_axes=[],
                current_batch_axes=current_batch_axes,
            )
           NotImplementedError: Differentiation rule for 'batchify' not implemented

...equinox/experimental/stateful.py:499: NotImplementedError
```
If this behaviour in differentiation is expected, with `FrozenBatchNorm` what I meant was, one can copy and re-define the running stats and weights/bias as `jnp.arrays` and allow *normal* computation of gradients without the overhead of tracking the state.

**Comment by patrick-kidger** [comment]
Hmm, that's not intended, and I think it should be fixable. (You're definitely putting the stateful/side-effecting operations through their paces.) Can you put together a MWE?

**Comment by paganpasta** [comment]
BatchNorm will fall in every aspect of CV! :D 

Cross-referencing older issue with [BN](https://github.com/patrick-kidger/equinox/issues/166).

As for the MWE, the net does not need to be deserialised etc. The error can be reproduced on a net after 1 round of training.

```python
import jax
import jax.numpy as jnp
import jax.random as jrandom
import optax

import numpy as np
import equinox as eqx


@eqx.filter_value_and_grad
def compute_loss(model, x, y, keys):
    logits = jax.vmap(model, axis_name=('batch'))(x, key=keys)
    one_hot_actual = jax.nn.one_hot(y, num_classes=5)
    return optax.softmax_cross_entropy(logits, one_hot_actual).mean()


@eqx.filter_jit
def make_step(model, x, y, keys, optimizer, opt_state):
    loss, grads = compute_loss(model, x, y, keys)
    updates, opt_state = optimizer.update(grads, opt_state)
    model = eqx.apply_updates(model, updates)
    return loss, model, opt_state


net = eqx.nn.Sequential(
    [
        eqx.nn.Linear(3, 5, key=jrandom.PRNGKey(0)),
        eqx.experimental.BatchNorm(5, axis_name='batch')
    ]
)

x = jnp.asarray(np.random.rand(10, 3))
y = jnp.asarray(np.random.randint(0, 9, 10))
key = jrandom.split(jrandom.PRNGKey(0), 10)

optimizer = optax.adam(learning_rate=0.1)
opt_state = optimizer.init(eqx.filter(net, eqx.is_array))
_, net, _ = make_step(net, x, y, key, optimizer, opt_state)
net = eqx.tree_inference(net, True)
# NotImplementedError
loss, grads = compute_loss(net, x, y, key)


```

**Comment by patrick-kidger** [comment]
Okay, I think this should now be fixed in #191.

Whilst looking into this, I realised that there is one other possible bug, in that one can still hit an assertion error if vmap'ing over an ensemble of BatchNorm-using models in inference mode.

For now I'm going to leave that one alone. Rationale:
- I'd need to think a lot harder about what the fix for that is.
- That's quite a niche scenario.
- It's a loud error; there shouldn't be any silent correctness issues.
- It's quite likely that everything in `eqx.experimental` is going to get deprecated in the near future anyway. (In favour of something done using proper stateful operations via JAX itself; as oft-commented-on by me...)

**Comment by paganpasta** [comment]
Thanks for the fix! 

I agree that revamping `experimental` given the introduction of stateful operations in the near future is not worth the effort. Moreover, I am sure majority of the use-cases are now handled by the current implementation.
